### PR TITLE
AK: fix overflow bug in binary_search

### PR DESCRIPTION
**Pull request by tryfinally**
fix for issue #2866 : binary search will fail on large arrays due to overflows and mixed signed/unsigned int operations.

### CODE REVIEW

File: AK/BinarySearch.h [hunk 1]
```diff
@@ -38,17 +38,23 @@ int integral_compare(const T& a, const T& b)
 }
 
 template<typename T, typename Compare>
-T* binary_search(T* haystack, size_t haystack_size, const T& needle, Compare compare = integral_compare, int* nearby_index = nullptr)
+T* binary_search(T* haystack, size_t haystack_size, const T& needle, Compare compare = integral_compare, size_t* nearby_index = nullptr)
 {
-    int low = 0;
-    int high = haystack_size - 1;
+    if(!haystack || !haystack_size)
```
Comment (awesomekling) [inline AK/BinarySearch.h:43]: Please use `clang-format`.
Also please add a test case to TestBinarySearch.cpp that fails without this fix :)

**Comment by tryfinally** [comment]
@awesomekling test added and clang-formated

**Comment by tryfinally** [comment]
build and test on Ubuntu 20 machine with 16GB memory: test pass with fixed code.
on CI test fails with bad_alloc

**Comment by awesomekling** [comment]
Ah, that's annoying. We obviously can't break the CI. Let's just put the test inside `#if 0` or something similar for now.

**Comment by tryfinally** [comment]
modified test: fails with old implementation and pass after this pull request.

**Comment by awesomekling** [comment]
I don't understand the state of this PR and #2876. This one has conflicts, and the other one breaks the build.
To merge, we need something that does not break the CI.

**Comment by tryfinally** [comment]
PR #2876 is not to be merged, it should be closed after this PR (#2837) is merged.

**Comment by tryfinally** [comment]
Regarding the  conflict:
0. No conflicts on my side
1. Have no idea how this came to be. 
2. How did the CI build passed on compile and tests? ( I doubled check the CI output)
3. No idea how to fix other than  opening a new PR

**Comment by tryfinally** [comment]
This PR was composed 14 days ago to fix AK:binary_search which has
been refactored by @asynts a week ago in PR #2884 
> Refactor: Change the AK::binary_search signature to use AK::Span.

I will close this PR and submit a new taking into account the Span param. See PR #2961

**Pull request closed without merging**